### ISSUE TITLE
Supernova applies gate to non-gated envelopes

### ISSUE THREAD

**Issue by jamshark70** [supercollider/supercollider#3094]
```
(
SynthDef(\test, { |out, t_trig = 1|
	var sig = SinOsc.ar(440, 0, 0.1),
	eg = EnvGen.kr(Env.perc(0.01, 1), t_trig, doneAction: 2);
	Out.ar(out, (sig * eg).dup);
}).add;
)

Synth(\test);
```

- scsynth: Correct, decaying tone.
- supernova: No sound. It cuts the envelope in the second control block.

**Comment by jamshark70** [comment]
If I had to guess -- it looks kind of like, in supernova, EnvGen with a timed envelope (releaseNode = -99) jumps to the release level when the gate input goes to 0 -- but, in scsynth, gate == 0 is ignored for timed envelopes (the correct behavior).

Or maybe gate == 0 is misinterpreted as an early release of 0 duration.

But I don't understand how this could happen, because EnvGen's source code is the same for both servers ... ??? I'm not imagining the discrepant behavior, though.

**Comment by patrickdupuis** [comment]
You're not imagining this. I can reproduce over here (linux).

**Comment by sonoro1234** [comment]
Only happens with t_trig not with trig

**Comment by sonoro1234** [comment]
Some observations:
scsynth printing check gate: prevGate, gate
and stage+1 > num stages means DoneAction is applied
```
check gate 0.000000,0.000000
stage 1000000000 rel -99 numstages 2
stage+1 > num stages
check gate 0.000000,1.000000
stage -1 rel -99 numstages 2
stage 0
check gate 1.000000,0.000000
...
```
but supernova prints
```
check gate 0.000000,0.000000
stage 1000000000 rel -99 numstages 2
stage+1 > num stages
check gate 0.000000,1.000000
stage -1 rel -99 numstages 2
stage 0
```
and stops printing. This means that supernova behaves as it should applying DoneAction when stage is 100000000 (ENVGEN_NOT_STARTED) and scsynth fails to apply DoneAction.

TrigControl give us check gate 0.000000,0.000000 instead of check gate 0.000000,1.000000 and is causing the problem

**Comment by brianlheim** [comment]
From #3650 - @sonoro1234 wrote:

---

this argument was lost in a closed thread:

Shouldn´t Ctor be?

```
void TrigControl_Ctor(Unit* unit)
{
	//Print("TrigControl_Ctor\n");
	if (unit->mNumOutputs == 1) {
		SETCALC(TrigControl_next_1);
		TrigControl_next_1(unit, 1);
	} else {
		SETCALC(TrigControl_next_k);
		TrigControl_next_k(unit, 1);
	}
	//ClearUnitOutputs(unit, 1);
}
```

This last change solves #3094

**Comment by jamshark70** [comment]
> EnvGen has a bug: no defined action when prevGate=0 and gate=0 just after construction so that stage =ENVGEN_NOT_STARTED goes to doneAction instead of just pause or remain at initial level. How should be done?

AFAIK scsynth's behavior in this case is to hold at the initial level until the gate input becomes positive. IMO, we "should" have a rule that scsynth's behavior is *de facto* correct, and drop-in server replacements should follow that standard as much as possible. Unfortunately, there are other cases in supernova where compatibility was not the first consideration (`/n_order` is one example, off the top of my head; higher strictness about argument types is another).

I had asked the question above (and I still don't understand) -- if the EnvGen source code is the same (there is no LFUGens-supernova-only.cpp), why is the initial gate behavior different between scsynth and supernova? EnvGen is deterministic. It seems pretty obviously wrong that the same source code, containing only deterministic operations and paying attention to thread safety, when compiled to two different targets, produces incompatible results.

Here, I'm out of my depth. I guess something in the build system is different for the `**_supernova` targets -- I had a quick look at server/plugins/CMakeLists.txt but it isn't obvious.

**Comment by sonoro1234** [comment]
@jamshark70 
See my second post.
Code is the same but in scsynth we see that stage+1 > num stages happens but DoneAction is not applied. This seems a bug in scsynth not being able to apply DoneAction for whatever reason that could be further investigated , more than a design behaviour.

**Comment by jamshark70** [comment]
If an Env has no release node, the correct behavior when `gate` falls to zero is to play through the envelope, all the way to the end.

My test case is about a timed, not gated, envelope. The correct behavior, then, is:

- If `gate` is 0 when the synth node starts, do not start the envelope -- hold at the initial level.

- Once `gate > 0`, play through the envelope to the end.

- If `gate < -1`, it may do an early release (linear envelope segment to 0, whose duration is `-1 - gate`. But, after the envelope starts playing, all other changes in `gate` should be ignored. `gate == 0` should *not* trigger any release behavior. It's impossible. Because the Env has no releaseNode, the release duration is unknown in this case.

Neither scsynth nor supernova should be applying the doneAction at all here.

**Comment by jamshark70** [comment]
Ah, OK, I reread the thread more carefully and I see where the misconception crept in.

> This means that supernova behaves as it should applying DoneAction when stage is 100000000 (ENVGEN_NOT_STARTED)

No, that isn't correct.

The DoneAction should apply only when EnvGen reaches the end of the envelope, never before. The catch is, EnvGen can't reach the end of the envelope if it doesn't move past the beginning.

If supernova "appl[ies] DoneAction when stage is 100000000 (ENVGEN_NOT_STARTED)," then supernova's behavior is incorrect. Scsynth is fine.

But I'm not sure how that applies to the test case. In my tests, in the first control block, `gate == 1`, falling to 0 in the second control block. At that point, the EnvGen *has* started; `stage` should be < numStages.

**Comment by sonoro1234** [comment]
here can be seen that doneAction is called in both servers when stage =ENVGEN_NOT_STARTED

https://github.com/supercollider/supercollider/blob/develop/server/plugins/LFUGens.cpp#L2826

**Comment by jamshark70** [comment]
> here can be seen that doneAction is called in both servers when stage =ENVGEN_NOT_STARTED

That `if` condition looks wrong to me. Based on a spec that makes any sort of sense at all, we should emphatically not execute the done action if the envelope hasn't started.

**Comment by jamshark70** [comment]
That is:

```
if (unit->m_stage+1 >= numstages)
```

Why isn't it this?

```
if (unit->m_stage+1 >= numstages && unit->m_stage < ENVGEN_NOT_STARTED)
```

Looks like someone at some point added ENVGEN_NOT_STARTED and forgot about this spot. Or maybe it was negative at some time in the past, but now it's positive. Doesn't really matter -- it's just that this branch doesn't make sense.

**Comment by sonoro1234** [comment]
Completely agree with that.
I will try your proposal.

**Comment by sonoro1234** [comment]
your proposal would make to take the third branch:
https://github.com/supercollider/supercollider/blob/develop/server/plugins/LFUGens.cpp#L2845

This issue was already addressed in:
https://github.com/supercollider/supercollider/commit/730070ad4a29b319dc05bdd3ec26d493eda3ee21
but introduced a new bug and was discarded in:
https://github.com/supercollider/supercollider/commit/0f698556cabac7884936208edbc100ec26926a88

So the solution would be to apply the first commit but make it  work on EnvGen_next_aa

**Comment by sonoro1234** [comment]
PR #3671 waiting for approval.

**Comment by brianlheim** [comment]
closed in #4436, thanks @sonoro1234 !
